jumanji/environments/logic/rubiks_cube/env.py

```diff
@@ -253,187 +211,49 @@ def action_spec(self) -> specs.MultiDiscreteArray:
-            - depth (0-cube_size//2). This indicates how many layers down from the face
-                the turn will take place.
-            - amount (0-2). This indicates the amount of turning (see below).
-
-        Convention:
-        - 0 = up face
-        - 1 = front face
-        - 2 = right face
-        - 3 = back face
-        - 4 = left face
-        - 5 = down face
-        All read in reading order when looking directly at a face.
-
-        To look directly at the faces:
-        - UP: LEFT face on the left and BACK face pointing up
-        - FRONT: LEFT face on the left and UP face pointing up
-        - RIGHT: FRONT face on the left and UP face pointing up
-        - BACK: RIGHT face on the left and UP face pointing up
-        - LEFT: BACK face on the left and UP face pointing up
-        - DOWN: LEFT face on the left and FRONT face pointing up
-
-        Turning amounts are when looking directly at a face:
-        - 0 = clockwise turn
-        - 1 = anticlockwise turn
-        - 2 = half turn
-        """
-        face_and_depth, amount = jnp.divmod(action, len(CubeMovementAmount))
-        face, depth = jnp.divmod(face_and_depth, self.cube_size // 2)
-        return jnp.stack([face, depth, amount], axis=0)
-
-    def _flatten_action(self, action: chex.Array) -> chex.Array:
-        """Inverse of the `_flatten_action` method."""
-        face, depth, amount = action
-        return (
-            face * len(CubeMovementAmount) * (self.cube_size // 2)
-            + depth * len(CubeMovementAmount)
-            + amount
-        )
-
-    def _rotate_cube(self, cube: Cube, flat_action: chex.Array) -> Cube:
-        """Apply a flattened action (index into the sequence of all moves) to a cube."""
-        moved_cube = jax.lax.switch(flat_action, self.all_moves, cube)
-        return moved_cube
-
-    def _scramble_solved_cube(self, flat_actions_in_scramble: chex.Array) -> Cube:
-        """Return a scrambled cube according to a given sequence of flat actions."""
-        cube = make_solved_cube(cube_size=self.cube_size)
-        cube, _ = jax.lax.scan(
-            lambda *args: (self._rotate_cube(*args), None),
-            cube,
-            flat_actions_in_scramble,
-        )
-        return cube
-
     def _state_to_observation(self, state: State) -> Observation:
         return Observation(cube=state.cube, step_count=state.step_count)
 
     def render(self, state: State) -> None:
-        """Render frames of the environment for a given state using matplotlib.
+        """Renders the current state of the cube.
 
         Args:
-            state: `State` object corresponding to the new state of the environment.
+            state: the current state to be rendered.
         """
-        self._clear_display()
-        fig, ax = self._get_fig_ax()
-        self._draw(ax, state)
-        self._update_display(fig)
+        self._env_viewer.render(state=state)
```

Comment:
In most environments, we try to return whatever the viewer returns so that in case the viewer is one that outputs a numpy array we have it. Please see e.g. TSP render.